core/node/node_framework/src/implementations/layers/state_keeper/output_handler.rs

```diff
@@ -93,6 +89,38 @@ impl OutputHandlerLayer {
         self.protective_reads_persistence_enabled = protective_reads_persistence_enabled;
         self
     }
+
+    async fn validate_l2_legacy_shared_bridge_addr(
+        &self,
+        pool: &ConnectionPool<Core>,
+    ) -> Result<(), WiringError> {
+        let mut connection = pool.connection().await.context("Get DB connection")?;
+
+        if let Some(l2_block) = connection
+            .blocks_dal()
+            .get_earliest_l2_block_number()
+            .await
+            .context("failed to load earliest l2 block number")?
+        {
+            let header = connection
+                .blocks_dal()
+                .get_l2_block_header(l2_block)
+                .await
+                .context("failed to load L2 block header")?
+                .context("missing L2 block header")?;
+            let protocol_version = header
+                .protocol_version
+                .unwrap_or_else(ProtocolVersionId::last_potentially_undefined);
+
+            if protocol_version.is_pre_gateway() && self.l2_legacy_shared_bridge_addr.is_none() {
```

Comment:
maybe also do the contrary: if it is post gateway and we provided shared bridge, fail also?

Reply:
IIUC it's not always correct. E.g. zk_toolbox currently init system with genesis_protocol_v = v25 and l2_legacy_shared_bridge_addr.is_some(), no?

Reply:
ah damn, this is not correct for production, only correct for local development